crates/block-producer/src/batch_builder/batch.rs

```diff
@@ -44,38 +47,61 @@ impl TransactionBatch {
     pub fn new(txs: Vec<ProvenTransaction>) -> Result<Self, BuildBatchError> {
         let id = Self::compute_id(&txs);
 
-        // TODO: we need to handle a possibility that a batch contains multiple transactions against
-        //       the same account (e.g., transaction `x` takes account from state `A` to `B` and
-        //       transaction `y` takes account from state `B` to `C`). These will need to be merged
-        //       into a single "update" `A` to `C`.
-        let updated_accounts =
-            txs.iter().map(|tx| (tx.id(), tx.account_update().clone())).collect();
+        let mut updated_accounts = vec![];
+        let mut produced_nullifiers = vec![];
+        let mut unauthenticated_input_notes = BTreeSet::new();
+        for tx in &txs {
+            // TODO: we need to handle a possibility that a batch contains multiple transactions against
+            //       the same account (e.g., transaction `x` takes account from state `A` to `B` and
+            //       transaction `y` takes account from state `B` to `C`). These will need to be merged
+            //       into a single "update" `A` to `C`.
+            updated_accounts.push((tx.id(), tx.account_update().clone()));
+
+            for note in tx.input_notes() {
+                produced_nullifiers.push(note.nullifier());
+                if let Some(header) = note.header() {
+                    if !unauthenticated_input_notes.insert(header.id()) {
+                        return Err(BuildBatchError::DuplicateUnauthenticatedNote(
+                            header.id(),
+                            txs,
+                        ));
+                    }
+                }
+            }
```

Comment:
With this logic, even if a note was produced and consumed in the same batch, we would still generate a nullifier for it. This means for all "ephemeral" notes we'll need to update the nullifier tree. It does simplify some things, but also may create quite a bit of an overhead. I'm still thinking through whether we should do something else here.